src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs

```diff
@@ -475,19 +475,23 @@ private void SendToDeadLetters(object message)
 
         private void PublishMessage(string path, object message, bool allButSelf = false)
         {
-            foreach (var entry in _registry)
+            IEnumerable<IActorRef> Refs()
             {
-                var address = entry.Key;
-                var bucket = entry.Value;
-
-                if (!(allButSelf && address == _cluster.SelfAddress) && bucket.Content.TryGetValue(path, out var valueHolder))
+                foreach (var entry in _registry)
                 {
-                    if (valueHolder != null && !valueHolder.Ref.Equals(ActorRefs.Nobody))
-                        valueHolder.Ref.Forward(message);
-                    else
-                        SendToDeadLetters(message);
+                    var address = entry.Key;
+                    var bucket = entry.Value;
+
+                    if (!(allButSelf && address == _cluster.SelfAddress) && bucket.Content.TryGetValue(path, out var valueHolder))
+                    {
+                        if (valueHolder != null && !Equals(valueHolder.Ref, ActorRefs.Nobody))
+                            yield return valueHolder.Ref;
+                    }
                 }
             }
+
+            if (!Refs().Any()) SendToDeadLetters(message);
```

Comment:
@ismaelhamed Have you tested performance here? LINQ could slow-down the code

Reply:
Nop, are there any performance benchmark in place that can be use?

Reply:
I guess you can at least be sure that yield generator will be called twice in most scenarios (1st time inside if statement, 2nd inside `Refs.ForEach`.

Reply:
@ismaelhamed no benchmarks for DPubSub at the moment, but best thing you could do here is to make sure that the list only gets evaluated once. Maybe just do a `foreach` loop over the `IEnumerable` and include a counter variable at the end. If the counter was not incremented, send it to dead letters.

Reply:
Not pretty though

```csharp
var counter = 0;
foreach (var r in Refs())
{
    if (r == null) continue;
    r.Forward(message);
    counter++;
}

if (counter == 0) SendToDeadLetters(message);
```

Reply:
@ismaelhamed high performance code often isn't :(